File: rust/src/generic/bit_packed_exponent.rs
use crate::generic::i_exponent::IExponent;
use std::hash::{Hash, Hasher};

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct BitPackedExponent {
    pub exponents: u64
}
impl BitPackedExponent {
    pub fn new(exponents: u64) -> Self {
        BitPackedExponent { exponents }
    }

    pub fn from_vec(exponents: [u8; 6]) -> Self {
        let mut packed: u64 = 0;
        
        // Pack the exponents with the first exponent at the left (bits 47..40)
        // Layout: [63..48]=degree, [47..40]=e0, [39..32]=e1, ... [7..0]=e5
        for (i, &exp) in exponents.iter().enumerate() {
            let shift = 40_u32.saturating_sub((8 * i) as u32);
            packed |= (exp as u64) << shift;
        }

        // Compute the total degree (sum of all exponents)
        let degree: u16 = exponents.iter().map(|&e| e as u16).sum();

        // Add the degree to the top 16 bits of the u64
        packed |= (degree as u64) << 48;

        BitPackedExponent { exponents: packed }
    }
}

impl Hash for BitPackedExponent {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.exponents.hash(state);
    }
}

impl IExponent for BitPackedExponent {
    fn add(&self, o: &Self) -> Self {
        BitPackedExponent { exponents: self.exponents + o.exponents }
    }

    fn sub(&self, o: &Self) -> Self {
        BitPackedExponent { exponents: self.exponents - o.exponents }
    }

    fn lcm(&self, other: &Self) -> Self {
        let self_exponents = self.exponents & 0x0000_FFFF_FFFF_FFFF;
        let other_exponents = other.exponents & 0x0000_FFFF_FFFF_FFFF;

        let mut lcm_exponents: u64 = 0;
        let mut degree: u16 = 0;

        for i in (0..48).step_by(8) {
            let self_exp = (self_exponents >> i) & 0xFF;
            let other_exp = (other_exponents >> i) & 0xFF;

            let lcm_exp = self_exp.max(other_exp);

            lcm_exponents |= lcm_exp << i;
            degree += lcm_exp as u16; // Accumulate the degree
        }

        // Set the degree in the top 16 bits
        lcm_exponents |= (degree as u64) << 48;

        BitPackedExponent { exponents: lcm_exponents }
    }

    fn degree(&self) -> u32 {
        ((self.exponents >> 48) & 0xFFFF) as u32
    }

    fn lex_compare(&self, other: &Self) -> std::cmp::Ordering {
        for i in (0..48).step_by(8).rev() {
            let self_exp = self.exponents & 0x0000_FFFF_FFFF_FFFF;
            let other_exp = other.exponents & 0x0000_FFFF_FFFF_FFFF;

            if self_exp < other_exp {
                return std::cmp::Ordering::Less;
            } else if self_exp > other_exp {
                return std::cmp::Ordering::Greater;
            }
        }
        std::cmp::Ordering::Equal
    }

    fn can_reduce(&self, divisor: &Self) -> bool {
        for i in (0..48).step_by(8) {
            let self_exp = (self.exponents >> i) & 0xFF;
            let divisor_exp = (divisor.exponents >> i) & 0xFF;

            if self_exp < divisor_exp {
                return false;
            }
        }
        true
    }

}

File: julia/generic/bit_packed_exponent.jl
import .IExponent

struct BitPackedExponent <: IExponent
    exps::UInt64
end

function BitPackedExponent(exponents::Vector{Int})
    packed = UInt64(0)
    padded_exps = copy(exponents)
    while length(padded_exps) < 6
        push!(padded_exps, 0)
    end
    
    for i in 1:6
        exp = padded_exps[i]
        shift = UInt64(40 - 8 * (i - 1))
        packed |= (UInt64(exp) & 0xFF) << shift
    end
    
    deg = UInt64(sum(padded_exps[1:6]))
    packed |= (deg & 0xFFFF) << 48
    
    return BitPackedExponent(packed)
end

function degree(a::BitPackedExponent)
    return (a.exps >> 48) & 0xFFFF
end

function lex_compare(a::BitPackedExponent, b::BitPackedExponent)
    self_exp = a.exps & 0x0000FFFFFFFFFFFF
    other_exp = b.exps & 0x0000FFFFFFFFFFFF
    return cmp(self_exp, other_exp)
end

function can_reduce(a::BitPackedExponent, b::BitPackedExponent)
    for i in 0:8:40
        self_exp = (a.exps >> i) & 0xFF
        divisor_exp = (b.exps >> i) & 0xFF
        if self_exp < divisor_exp
            return false
        end
    end
    return true
end

function get_lcm(a::BitPackedExponent, b::BitPackedExponent)
    self_exponents = a.exps & 0x0000FFFFFFFFFFFF
    other_exponents = b.exps & 0x0000FFFFFFFFFFFF
    
    lcm_exponents = UInt64(0)
    deg = UInt64(0)
    
    for i in 0:8:40
        self_exp = (self_exponents >> i) & 0xFF
        other_exp = (other_exponents >> i) & 0xFF
        lcm_exp = max(self_exp, other_exp)
        lcm_exponents |= (lcm_exp << i)
        deg += lcm_exp
    end
    
    lcm_exponents |= (deg & 0xFFFF) << 48
    return BitPackedExponent(lcm_exponents)
end


function Base.show(io::IO, a::BitPackedExponent)
    deg = (a.exps >> 48) & 0xFFFF
    print(io, "Degree: ", string(deg, base=16, pad=4), ", Exponents (hex): ")
    for i in 40:-8:0
        exp = (a.exps >> i) & 0xFF
        print(io, string(exp, base=16, pad=2), " ")
    end
end

Base.:+(a::BitPackedExponent, b::BitPackedExponent) = BitPackedExponent(a.exps + b.exps)
Base.:-(a::BitPackedExponent, b::BitPackedExponent) = BitPackedExponent(a.exps - b.exps)
Base.:(==)(a::BitPackedExponent, b::BitPackedExponent) = a.exps == b.exps
Base.hash(a::BitPackedExponent, h::UInt) = hash(a.exps, h)
